Libraries/LibWeb/WebAssembly/Memory.cpp

```diff
@@ -80,32 +95,42 @@ WebIDL::ExceptionOr<u32> Memory::grow(u32 delta)
     if (!memory->grow(delta * Wasm::Constants::page_size, Wasm::MemoryInstance::GrowType::No, Wasm::MemoryInstance::InhibitGrowCallback::Yes))
         return vm.throw_completion<JS::RangeError>("Memory.grow() grows past the stated limit of the memory instance"sv);
 
-    TRY(reset_the_memory_buffer());
+    TRY(refresh_the_memory_buffer(vm, realm(), m_address));
 
     return previous_size;
 }
 
 // https://webassembly.github.io/spec/js-api/#refresh-the-memory-buffer
-// FIXME: `refresh-the-memory-buffer` is a global abstract operation.
-//        Implement it as a static function to align with the spec.
-WebIDL::ExceptionOr<void> Memory::reset_the_memory_buffer()
+WebIDL::ExceptionOr<void> Memory::refresh_the_memory_buffer(JS::VM& vm, JS::Realm& realm, Wasm::MemoryAddress address)
 {
-    if (!m_buffer)
-        return {};
+    // 1. Let map be the surrounding agent’s associated Memory object cache.
+    // 2. Assert: map[memaddr] exists.
+    // 3. Let memory be map[memaddr].
+    auto& cache = Detail::get_cache(realm);
+    auto maybe_memory = cache.get_memory_instance(address);
+    VERIFY(maybe_memory.has_value());
+    auto memory = maybe_memory.release_value();
 
-    auto& vm = this->vm();
-    auto& realm = *vm.current_realm();
+    // 4. Let buffer be memory.[[BufferObject]].
+    auto& buffer = memory->m_buffer;
 
-    if (m_buffer->is_fixed_length()) {
+    if (!buffer)
+        return {};
+
+    // 5. If IsFixedLengthArrayBuffer(buffer) is true,
+    if (buffer->is_fixed_length()) {
         // https://webassembly.github.io/threads/js-api/index.html#refresh-the-memory-buffer
         // 1. If IsSharedArrayBuffer(buffer) is false,
-        if (!m_buffer->is_shared_array_buffer()) {
+        if (!buffer->is_shared_array_buffer()) {
             // 1. Perform ! DetachArrayBuffer(buffer, "WebAssembly.Memory").
-            MUST(JS::detach_array_buffer(vm, *m_buffer, JS::PrimitiveString::create(vm, "WebAssembly.Memory"_string)));
+            MUST(JS::detach_array_buffer(vm, *buffer, JS::PrimitiveString::create(vm, "WebAssembly.Memory"_string)));
         }
+
+        // 2. Let newBuffer be the result of creating a fixed length memory buffer from memaddr.
+        // 3. Set memory.[[BufferObject]] to newBuffer.
     }
 
-    m_buffer = TRY(create_a_fixed_length_memory_buffer(vm, realm, m_address, m_shared));
+    buffer = TRY(create_a_fixed_length_memory_buffer(vm, realm, address, memory->m_shared));
```

Comment:
should the assignment & co be moved into the if? the spec steps you have seem to suggest so at least.